File: mnmcp-v2/src/miniworld/protocol.py
import struct
from dataclasses import dataclass
from typing import Union

@dataclass
class LoginPacket:
    uin: int
    token: str

@dataclass
class RoomListPacket:
    pass

@dataclass
class JoinRoomPacket:
    room_id: str

Packet = Union[LoginPacket, RoomListPacket, JoinRoomPacket]

class Packet:
    @staticmethod
    def serialize(packet: Packet) -> bytes:
        if isinstance(packet, LoginPacket):
            return struct.pack("!BQ", 0x01, packet.uin) + packet.token.encode()
        elif isinstance(packet, RoomListPacket):
            return struct.pack("!B", 0x02)
        elif isinstance(packet, JoinRoomPacket):
            return struct.pack("!B", 0x03) + packet.room_id.encode()
        return b""
    
    @staticmethod
    def deserialize(data: bytes) -> Optional[Packet]:
        if len(data) < 1:
            return None
        packet_type = data[0]
        if packet_type == 0x01:
            uin = struct.unpack("!Q", data[1:9])[0]
            token = data[9:].decode()
            return LoginPacket(uin=uin, token=token)
        elif packet_type == 0x02:
            return RoomListPacket()
        elif packet_type == 0x03:
            room_id = data[1:].decode()
            return JoinRoomPacket(room_id=room_id)
        return None

File: src/rust/mnmcp-core/src/protocol.rs
//! Protocol module for MnMCP
//! 
//! MiniWorld and Minecraft protocol implementations

use serde::{Serialize, Deserialize};

/// MiniWorld packet types
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum MiniWorldPacket {
    Login(LoginPacket),
    RoomList(RoomListPacket),
    JoinRoom(JoinRoomPacket),
    GameData(GameDataPacket),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LoginPacket {
    pub uin: String,
    pub token: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RoomListPacket {
    pub rooms: Vec<RoomInfo>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RoomInfo {
    pub room_id: String,
    pub room_name: String,
    pub host_name: String,
    pub current_players: u32,
    pub max_players: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JoinRoomPacket {
    pub room_id: String,
    pub player_uin: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GameDataPacket {
    pub data: Vec<u8>,
}

/// Parse MiniWorld packet from bytes
pub fn parse_packet(data: &[u8]) -> Result<MiniWorldPacket, &'static str> {
    // TODO: Implement actual parsing
    Err("Not implemented")
}
